src/planscape/forsys/parse_forsys_output.py

```diff
@@ -0,0 +1,160 @@
+import rpy2
+
+import numpy as np
+
+from typing import TypedDict
+
+
+class RankedProject(TypedDict):
+    # Project ID.
+    id: int
+    # Contribution of each priority to the total score
+    # The contribution is weighted according to a scenario's priority weights.
+    weighted_priority_scores: dict
+    # The total score, summed across weighted priority scores.
+    total_score: float
+    # Project rank.
+    rank: int
+
+
+class Scenario(TypedDict):
+    # Priority weights for the scenario.
+    priority_weights: dict
+    # A list of the projects, ranked.
+    ranked_projects: list[RankedProject]
+    # Given ranked projects, a cumulative sum of project area.
+    cumulative_ranked_project_area: list[float]
+    # Given ranked projects, a cumulative sum of project cost.
+    cumulative_ranked_project_cost: list[float]
+
+
+# Transforms the output of a Forsys scenario set run into a more easily-interpreted version.
+class ForsysScenarioSetOutput():
+    scenarios: dict
+
+    # Converts a ForsysScenarioSetOutput to a dictionary.
+    # This facilitates JSON conversion.
+    def to_dictionary(self) -> dict:
+        output = {}
+        for k in self.scenarios.keys():
+            output[k] = self.scenarios[k]
+        return output
+
+    # Initializes a ForsysScenarioSetOutput instance given raw forsys output, and the following inputs to the forsys call: header names, list of priorities.
+    # Of note, priorities must be listed in the same order they're listed for the forsys call.
+    def __init__(self, raw_forsys_output: "rpy2.robjects.vectors.ListVector", priorities: list[str], project_id_header: str, area_header: str, cost_header: str):
+        self.scenarios = {}
+
+        self.__save_raw_forsys_output_as_dict(raw_forsys_output)
+
+        self.__set_header_names(priorities, area_header,
+                                cost_header, project_id_header)
+
+        for i in range(len(self.__forsys_output_dict[project_id_header])):
+            scenario_weights, scenario_str = self.__get_scenario(i)
+
+            if scenario_str in self.scenarios.keys():
+                self.__append_project_to_existing_scenario(
+                    scenario_str, scenario_weights, i)
+            else:
+                self.__append_project_to_new_scenario(
+                    scenario_str, scenario_weights, i)
+
+    # The raw forsys output consists of 3 R dataframes. This is the index of the "project output" dataframe.
+    __PROJECT_OUTPUT_INDEX = 1
+    # The priority weight header in the "project output" dataframe.
```

Comment:
I think this would be easier to read if

1. The constants were at the top of the class
2. The declarations of types for instance variables below them
3. The __init__ method next

Reply:
done - I'll use this pattern for other classes, too.